[SATURN-1884] Uploader page

andrewmarcus opened the pull request:
This PR implements a new feature allowing state labs to easily upload their entity data and create new tables in their workspace. Much of this functionality should eventually be ported back into the Workspaces view.

Amongst the recent updates:

* The abort upload button now works. I see that it can sometimes get into a state where the upload has stopped but the modal isn't dismissable, which I think happens because when the actual fetch is aborted, an unresolved promise is returned. But I didn't want to change the plumbing of the deep-down AJAX code. In the general case the button is working.
* I've switched out the UI to more of an accordion style rather than the nav buttons on the left side. After putting it in place, this interface seems MUCH more fluid and intuitive. See what you think.

I have not yet managed to resolve the issue with the `AutoSizer`. Please have your UX team take a look at it and tell me what I did wrong.

andrewmarcus committed:
Initial build-out of the new Uploader page

andrewmarcus committed:
Moved the uploader modal into a reusable component and encapsulated its state, added a new step for selecting or creating a folder in the uploads directory

andrewmarcus committed:
Added Metadata Upload functionality, updated progress bar uploader design

andrewmarcus committed:
Updated AutoSizer container to try to fix sizing issues

andrewmarcus committed:
Added final metadata upload step

andrewmarcus committed:
Store uploader's workspace and collection in the URL instead of state history, fixed logic bugs in step state synchronization

andrewmarcus committed:
Cleaned up a bunch of linter errors
More progress on upload aborter

andrewmarcus committed:
Cleaned up a bunch of linter errors
More progress on upload aborter

andrewmarcus committed:
Split the file browser/uploader into its own component to allow for reuse

andrewmarcus committed:
Remove the uploads page from the main menu

andrewmarcus committed:
Switched to using an accordion menu to display previous state, which significantly simplifies the usability of the interface

andrewmarcus committed:
Switched to using an accordion menu to display previous state, which significantly simplifies the usability of the interface

andrewmarcus committed:
Accessibility enhancements:
* Allow spacebar to select buttons (generally)
* Added aria-modal='true' to modals (generally)
* Show hover states on focus (generally)
* Move focus to section header when it opens, to help the user track their place (Uploads)
* Read out the current filter state to screen readers (Uploads > workspace)

andrewmarcus commented:
I just got some accessibility improvements in to the Uploader feature, some of which improve the site generally.

* Allow spacebar to select buttons (generally)
* Added aria-modal='true' to modals (generally)
* Show hover states on focus (generally)
* Move focus to section header when it opens, to help the user track their place (Uploads)
* Read out the current filter state to screen readers (Uploads > workspace)

I also hid the button for creating new folders for new

andrewmarcus committed:
Improved a11y

andrewmarcus committed:
Added a live region in the Uploader progress modal to inform screenreader users as file uploading progresses

dmohs commented on src/components/Dropzone.js:
```diff
@@ -25,10 +30,18 @@ const Dropzone = ({ disabled = false, onDragOver, onDrop, onDragLeave, style = {
   })
 
-  return div(getRootProps({ tabIndex: -1, style: { ...style, ...(dragging ? activeStyle : {}) } }), [
+  return div(getRootProps({ tabIndex: -1, style: { ...style, ...(dragging ? activeStyle : {}), ...(rejected ? rejectedStyle : {}) } }), [
```
Not seeing usage of `rejectedStyle` or `fileRejections` anywhere.

andrewmarcus replied:
Thanks for the feedback. I'd be happy to pull out the more general accessibility changes into separate PRs. There are a couple of things in here, such as this change to Dropzone that I was going to use but then ended up doing something else. They might still be useful in the future but I'll pull them out for this pass.

dmohs commented on src/components/Interactive.js:
```diff
@@ -46,7 +46,7 @@ const Interactive = Utils.forwardRefWithName('Interactive', ({
     style: { ...style, ...cssVariables, cursor: computedCursor, outline },
     onKeyDown: onKeyDown || (e => {
-      if (e.key === 'Enter') {
+      if (e.key === 'Enter' || e.key === ' ') {
```
I think this makes sense, but should be a separate PR unless essential.

dmohs commented on src/components/Modal.js:
```diff
@@ -53,7 +53,7 @@ const Modal = ({ onDismiss, title, titleExtras, children, width = 450, showCance
     },
     style: { overlay: styles.overlay, content: { ...styles.modal, width } },
-    aria: { labelledby: titleId },
+    aria: { labelledby: titleId, modal: true },
```
This is a great change, but also a separate PR unless essential to this functionality.

andrewmarcus replied:
Sure. I'll pull this and the other changes out into a separate PR handling the first pass at accessibility improvements. These particular roles and keyboard actions impact the accessiblity of the new uploads page, which is why I included them, but there are no particular dependencies that these be in there for the first pass.

dmohs commented on src/components/Modal.js:
```diff
@@ -74,9 +74,9 @@ const Modal = ({ onDismiss, title, titleExtras, children, width = 450, showCance
-        [_.isFunction(okButton), () => h(ButtonPrimary, { onClick: okButton }, 'OK')],
+        [okButton === undefined, () => h(ButtonPrimary, { onClick: onDismiss, danger }, 'OK')],
+        [_.isString(okButton), () => h(ButtonPrimary, { onClick: onDismiss, danger }, okButton)],
+        [_.isFunction(okButton), () => h(ButtonPrimary, { onClick: okButton, danger }, 'OK')],
```
I like this, but consider whether it can be omitted for the sake of preserving existing files as much as possible (when introducing new functionality). This would be a great addition as a follow-up PR.

andrewmarcus replied:
This particular change allowed me to make the Abort Upload button red instead of blue, which more accurately captures its purpose (don't press that button unless you really mean it). However, I'd be happy to pull that out into a different PR for cleanliness.

dmohs replied:
Tough call, probably should stay in this PR.

dmohs commented on src/components/table.js:
```diff
@@ -394,25 +394,35 @@ GridTable.propTypes = {
           return div({
             key,
             className: 'table-cell',
+            role: 'gridcell',
```
Love these role additions, but as a separate PR so we can contain the new functionality as much as possible.

dmohs commented on src/style.css:
```diff
@@ -7,7 +7,7 @@
 @import-normalize;
 
-.hover-style:hover {
+.hover-style:hover, .hover-style:focus {
```
Separate PR? I think we want these kinds of changes unconditionally, whereas the upload feature has some organizational controversy around it.

andrewmarcus replied:
Thanks for that, good to know. Happy to split it up

dmohs reviewed (commented):
This first round is focused on keeping the changed set contained as much as possible. In general, I'm impressed with how well you follow the existing style.

dmohs commented on src/components/FloatingActionButton.js:
```diff
@@ -23,7 +23,8 @@ const FloatingActionButton = ({ label, iconShape, onClick, bottom = 30, right =
         setHover(false)
-      }
+      },
+      ...props
```
This isn't right, I think. The ref you're passing here is creating a warning. I believe you should be able to find the right pattern elsewhere in the codebase.

andrewmarcus replied:
In the FloatingActionButton? I'm not sure I understand which ref you're referring to. The only thing I changed in this file was to pass through additional optional properties to the button. That said, I don't recall now whether I actually ended up needing the extra props here. I'll double-check and remove them if I can.

dmohs replied:
It's used to pass a ref on line 683 of Upload.js. I'd say leave that out until we can get it right. We can do a usability pass later.

dmohs replied:
(the ref is used to focus a button)

andrewmarcus committed:
Merge branch 'dev' into uploader

andrewmarcus committed:
Rolled back changes for features that didn't end up getting implemented, per PR feedback.
Rolled back a11y upgrades for inclusion in a separate PR
Additional work on Autosizer bug, better but not quite right yet

andrewmarcus committed:
Rolled back a11y upgrades for inclusion in a separate PR

andrewmarcus committed:
Fixed the table AutoSizer by fixing all the flexboxes to pass down 100% height everywhere

andrewmarcus commented:
I got the table AutoSizer working. The problem turned out to be that I just hadn't been passing down the `height: 100%` appropriately through all the flexbox layers.

andrewmarcus committed:
Merge branch 'dev' into uploader

dmohs commented:
This is excellent. Changes are well-contained to new files, additions to existing files, and very few changes to existing code, where the existing style is respected. I judge this is safe to merge and we can follow-up with improvements in later PRs. I will work on getting a second approval.

dmohs commented:
Tested manually by running through the demonstrated workflow. Only minor cosmetic and usability issues found, nothing that should block release.

abaumann reviewed (approved):
Reviewed the functionality and intended flow by running this branch locally and it all looks correct to me according to the requirements
